File: src/main/java/com/heybys/oddments/fooddelivery/domain/generic/TimePeriod.java
package com.heybys.oddments.fooddelivery.domain.generic;

import java.time.LocalTime;

import jakarta.persistence.Embeddable;

import com.heybys.oddments.base.domain.ValueObject;

import lombok.Getter;

@Getter
@Embeddable
public class TimePeriod extends ValueObject<TimePeriod> {

    private LocalTime startTime;
    private LocalTime endTime;

    public static TimePeriod between(LocalTime startTime, LocalTime endTime) {
        return new TimePeriod(startTime, endTime);
    }

    public TimePeriod() {}

    public TimePeriod(LocalTime startTime, LocalTime endTime) {
        this.startTime = startTime;
        this.endTime = endTime;
    }

    @Override
    public boolean equals(Object other) {
        return super.equals(other);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    public boolean contains(LocalTime datetime) {
        return (datetime.isAfter(startTime) || datetime.equals(startTime))
                && (datetime.isBefore(endTime) || datetime.equals(endTime));
    }

    public TimePeriod putOffHours(int hours) {
        return new TimePeriod(startTime.plusHours(hours), endTime.plusHours(hours));
    }

    public TimePeriod putOffHours(long hours) {
        return new TimePeriod(startTime.plusHours(hours), endTime.plusHours(hours));
    }
}

File: src/main/kotlin/com/heybys/oddments/fooddelivery/domain/generic/TimePeriod.kt
package com.heybys.oddments.fooddelivery.domain.generic

import com.heybys.oddments.base.domain.ValueObject
import jakarta.persistence.Embeddable
import java.time.LocalTime

@Embeddable
class TimePeriod(val startTime: LocalTime? = null, val endTime: LocalTime? = null) :
    ValueObject<TimePeriod>() {

    // JPA constructor
    constructor() : this(null, null)

    companion object {
        @JvmStatic
        fun between(startTime: LocalTime, endTime: LocalTime): TimePeriod =
            TimePeriod(startTime, endTime)
    }

    override fun equals(other: Any?): Boolean = super.equals(other)

    override fun hashCode(): Int = super.hashCode()

    fun contains(datetime: LocalTime): Boolean {
        return startTime != null &&
            endTime != null &&
            (datetime.isAfter(startTime) || datetime == startTime) &&
            (datetime.isBefore(endTime) || datetime == endTime)
    }

    fun putOffHours(hours: Int): TimePeriod {
        return if (startTime != null && endTime != null) {
            TimePeriod(startTime.plusHours(hours.toLong()), endTime.plusHours(hours.toLong()))
        } else {
            this
        }
    }

    fun putOffHours(hours: Long): TimePeriod {
        return if (startTime != null && endTime != null) {
            TimePeriod(startTime.plusHours(hours), endTime.plusHours(hours))
        } else {
            this
        }
    }
}
